### PR TITLE
[#573] BackupRequest doesn't work with streaming

### PR DESCRIPTION
**Pull request by zizhong**
Details can be found in #573.
The fix is to buffer the body and resend if needed.

cc @nizarm @ssheng

### CODE REVIEW

File: d2/src/main/java/com/linkedin/d2/balancer/clients/BackupRequestsClient.java [hunk 1]
```diff
@@ -380,6 +384,29 @@ public void streamRequest(StreamRequest request, Callback<StreamResponse> callba
   @Override
   public void streamRequest(StreamRequest request, RequestContext requestContext, Callback<StreamResponse> callback)
   {
+    // Buffering stream request raises concerns on memory usage and performance.
+    // Currently only support backup requests with IS_FULL_REQUEST.
+    if (!isFullRequest(requestContext)) {
```
[the remark this answers is not in the document]
Comment (zizhong) [reply d2/src/main/java/com/linkedin/d2/balancer/clients/BackupRequestsClient.java:389]: @nizarm Yes, we require them to set the flag if they want to use backup request with streaming.

[the remark this answers is not in the document]
Comment (nizarm) [reply d2/src/main/java/com/linkedin/d2/balancer/clients/BackupRequestsClient.java:389]: Sounds good

**Review by nizarm — COMMENTED** [review]
Can you check why the travis build is failing?

**Comment by zizhong** [comment]
@nizarm I figured we need to use the same copy of RequestContext when sending request so that the targetHostHint can work.

**Review by nizarm — CHANGES_REQUESTED** [review]

File: d2/src/main/java/com/linkedin/d2/balancer/clients/BackupRequestsClient.java [hunk 2]
```diff
@@ -532,9 +560,23 @@ private void maybeSendBackupRequest()
...
+          if (_request instanceof StreamRequest) {
+            StreamRequest req = (StreamRequest)_request;
+            req = req.builder()
+                .build(EntityStreams.newEntityStream(new ByteStringWriter(
+                    (ByteString) _requestContext.getLocalAttr(R2Constants.BACKUP_REQUEST_BUFFERED_BODY)
+                )));
+            request = (R)req;
+            needCloneRC = true;
+          }
+          _client.doRequest(request, needCloneRC ? _requestContext.clone(): _requestContext, new Callback<T>()
```
[the remark this answers is not in the document]
Comment (zizhong) [reply d2/src/main/java/com/linkedin/d2/balancer/clients/BackupRequestsClient.java]: @nizarm Sorry, but I think you are reviewing an outdated diff. Please check the latest. Thanks!

**Review by nizarm — APPROVED** [review]

**Comment by nizarm** [comment]
Nice work, thanks for taking care of this usecase

**Pull request merged**